include/aws/io/retry_strategy.h

```diff
@@ -0,0 +1,178 @@
+ * aws_retry_strategy_release_retry_token().
+ */
+typedef void(aws_retry_strategy_on_retry_ready_fn)(struct aws_retry_token *token, int error_code, void *user_data);
+
+enum aws_retry_error_type {
+    /** This is a connection level error such as a socket timeout, socket connect error, tls negotiation timeout etc...
+     * Typically these should never be applied for non-idempotent request types since in this scenario, it's impossible
+     * to know whether the operation had a side effect on the server. */
+    AWS_RETRY_ERROR_TYPE_TRANSIENT,
+    /** This is an error where the server explicitly told the client to back off, such as a 429 or 503 Http error. */
+    AWS_RETRY_ERROR_TYPE_THROTTLING,
+    /** This is a server error that isn't explicitly throttling but is considered by the client
+     * to be something that should be retried. */
+    AWS_RETRY_ERROR_TYPE_SERVER_ERROR,
+    /** Doesn't count against any budgets. This could be something like a 401 challenge in Http. */
+    AWS_RETRY_ERROR_TYPE_CLIENT_ERROR,
+};
+
+struct aws_retry_strategy_vtable {
+    void (*destroy)(struct aws_retry_strategy *retry_strategy);
+    int (*acquire_token)(
+        struct aws_retry_strategy *retry_strategy,
+        const struct aws_byte_cursor *partition_id,
+        aws_retry_strategy_on_retry_token_acquired_fn *on_acquired,
+        void *user_data,
+        uint64_t timeout_ms);
+    int (*schedule_retry)(
+        struct aws_retry_token *token,
+        enum aws_retry_error_type error_type,
+        aws_retry_strategy_on_retry_ready_fn *retry_ready,
+        void *user_data);
+    int (*record_success)(struct aws_retry_token *token);
+    void (*release_token)(struct aws_retry_token *token);
+};
+
+struct aws_retry_strategy {
+    struct aws_allocator *allocator;
+    struct aws_retry_strategy_vtable *vtable;
+    struct aws_atomic_var ref_count;
+    void *impl;
+};
+
+struct aws_retry_token {
+    struct aws_allocator *allocator;
+    struct aws_retry_strategy *retry_strategy;
+    void *impl;
+};
+
+/**
+ * Jitter mode for exponential backoff.
+ *
+ * For a great writeup on these options see:
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+ */
+enum aws_exponential_backoff_jitter_mode {
+    /* Uses AWS_EXPONENTIAL_BACKOFF_JITTER_FULL */
+    AWS_EXPONENTIAL_BACKOFF_JITTER_DEFAULT,
+    AWS_EXPONENTIAL_BACKOFF_JITTER_NONE,
+    AWS_EXPONENTIAL_BACKOFF_JITTER_FULL,
+    AWS_EXPONENTIAL_BACKOFF_JITTER_DECORRELATED,
+};
+
+/**
+ * Options for exponential backoff retry strategy. el_group must be set, any other option, if set to 0 will signify
+ * "use defaults"
+ */
+struct aws_exponential_backoff_retry_options {
+    /** Event loop group to use for scheduling tasks. */
+    struct aws_event_loop_group *el_group;
+    /** Max retries to allow. The default value is 10 */
+    size_t max_retries;
+    /** Scaling factor to add for the backoff. Default is 25ms */
+    uint32_t backoff_scale_factor_ms;
```

Comment:
[just thinking out loud]: features we might want in some far-flung future:
1) unlimited retries
2) capped backoff time

(ex: MQTT's reconnect logic)